File: module4_dr_grading/src/utils/preprocessImage.m
function img = preprocessImage(img)
% PREPROCESSIMAGE Crops black borders, pads to a square to preserve aspect
% ratio, resizes to 224x224, and normalizes using ImageNet statistics.
%
% Inputs:
%   img - RGB image (uint8 or double, any size)
% Outputs:
%   img - Preprocessed single-precision 224x224x3 image tensor

    % Convert input image to double in [0, 1] range
    img = double(img);
    if max(img(:)) > 1.0
        img = img / 255.0;
    end

    % Ensure 3 color channels
    [h, w, c] = size(img);
    if c == 1
        img = cat(3, img, img, img);
    elseif c > 3
        img = img(:, :, 1:3);
    end

    % 1. Crop black margins
    % Convert to grayscale for thresholding
    gray = img(:, :, 1) * 0.2989 + img(:, :, 2) * 0.5870 + img(:, :, 3) * 0.1140;
    mask = gray > 0.05;
    [rows, cols] = find(mask);
    
    if ~isempty(rows) && ~isempty(cols)
        minRow = min(rows);
        maxRow = max(rows);
        minCol = min(cols);
        maxCol = max(cols);
        img = img(minRow:maxRow, minCol:maxCol, :);
    end

    % 2. Pad to square to preserve aspect ratio
    [h, w, ~] = size(img);
    if h > w
        padTotal = h - w;
        padLeft = floor(padTotal / 2);
        padRight = padTotal - padLeft;
        img = padarray(img, [0, padLeft, 0], 0, 'pre');
        img = padarray(img, [0, padRight, 0], 0, 'post');
    elseif w > h
        padTotal = w - h;
        padTop = floor(padTotal / 2);
        padBottom = padTotal - padTop;
        img = padarray(img, [padTop, 0, 0], 0, 'pre');
        img = padarray(img, [padBottom, 0, 0], 0, 'post');
    end

    % 3. Resize to 224x224
    img = imresize(img, [224, 224]);

    % 4. Normalization (ImageNet mean and std dev)
    meanVal = reshape([0.485, 0.456, 0.406], [1, 1, 3]);
    stdVal = reshape([0.229, 0.224, 0.225], [1, 1, 3]);
    img = (img - meanVal) ./ stdVal;

    % 5. Convert to single precision
    img = single(img);
end

File: module4_dr_grading python/src/utils/preprocess_image.py
import numpy as np
import cv2

def preprocess_image(img):
    """
    Crops black borders, pads to a square to preserve aspect
    ratio, resizes to 224x224, and normalizes using ImageNet statistics.
    
    Inputs:
      img - RGB image (uint8 or float, any size)
    Outputs:
      img_tensor - Preprocessed float32 224x224x3 image array 
                   (Can be converted to PyTorch tensor [3, 224, 224] later)
    """
    # Convert input image to double in [0, 1] range
    img_double = img.astype(float)
    if np.max(img_double) > 1.0:
        img_double = img_double / 255.0
        
    # Ensure 3 color channels
    if len(img_double.shape) == 2:
        img_double = np.stack([img_double, img_double, img_double], axis=-1)
    elif img_double.shape[2] > 3:
        img_double = img_double[:, :, :3]
        
    # 1. Crop black margins
    # Convert to grayscale for thresholding
    gray = img_double[:, :, 0] * 0.2989 + img_double[:, :, 1] * 0.5870 + img_double[:, :, 2] * 0.1140
    mask = gray > 0.05
    
    rows, cols = np.where(mask)
    
    if len(rows) > 0 and len(cols) > 0:
        min_row, max_row = np.min(rows), np.max(rows)
        min_col, max_col = np.min(cols), np.max(cols)
        img_double = img_double[min_row:max_row+1, min_col:max_col+1, :]
        
    # 2. Pad to square to preserve aspect ratio
    h, w, c = img_double.shape
    if h > w:
        pad_total = h - w
        pad_left = pad_total // 2
        pad_right = pad_total - pad_left
        img_double = np.pad(img_double, ((0, 0), (pad_left, pad_right), (0, 0)), mode='constant', constant_values=0)
    elif w > h:
        pad_total = w - h
        pad_top = pad_total // 2
        pad_bottom = pad_total - pad_top
        img_double = np.pad(img_double, ((pad_top, pad_bottom), (0, 0), (0, 0)), mode='constant', constant_values=0)
        
    # 3. Resize to 224x224
    img_resized = cv2.resize(img_double, (224, 224), interpolation=cv2.INTER_LINEAR)
    
    # 4. Normalization (ImageNet mean and std dev)
    mean_val = np.array([0.485, 0.456, 0.406]).reshape(1, 1, 3)
    std_val = np.array([0.229, 0.224, 0.225]).reshape(1, 1, 3)
    img_norm = (img_resized - mean_val) / std_val
    
    # 5. Convert to single precision
    return img_norm.astype(np.float32)
